source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,140 @@
+#include "common/http/filter/cors_filter.h"
+
+#include "envoy/http/codes.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Http {
+
+CorsFilter::CorsFilter(CorsFilterConfigConstSharedPtr config)
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  initialize();
+
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  const Http::HeaderEntry* method = headers.Method();
+  if (method != nullptr && method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!cors_enabled_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_origin_ == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
```

Comment:
@codesuki check here if the input Origin value matches the one in the config, no?

Reply:
something like:
```c++
if (allow_origin_.find(*origin_) == std::string::npos) {
  return FilterHeadersStatus::Continue;
}

response_headers->insertAccessControlAllowOrigin().value(*origin_);

if (allow_credentials_) {
  response_headers->insertAccessControlAllowCredentials().value(
      Http::Headers::get().CORSValues.True);
}
```

Reply:
Thanks! Your comment made me re-read the RFC. I first thought we can use regexes but doesn't seem so after all. I'll read the spec once more and make some changes.

Reply:
Per [w3c](https://www.w3.org/TR/cors/#access-control-allow-origin-response-hea) the header must contain the domain that was sent as the Origin header for the request to be accepted.  How the server decides how to accept that string is implementation dependent.